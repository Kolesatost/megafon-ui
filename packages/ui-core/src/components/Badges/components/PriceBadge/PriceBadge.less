@import (reference) '~styles/base.less';

@b: e('.mfui-price-badge');

@{b} {
    .smallFont();

    display: flex;
    align-items: center;

    padding: 4px;
    padding-right: 8px;
    width: fit-content;
    height: 24px;

    border-radius: 8px;

    &_adaptive {
        @media @desktop {
            .commonFont();

            padding: 4px 8px;
            height: 32px;

            border-radius: 12px;
        }
    }
    
    &_theme_grey {
        background-color: var(--spbSky0);

        @{b}__text {
            color: var(--spbSky3);
        }

        @{b}__icon {
            fill: var(--spbSky3);
        }
    }

    &_theme_orange {
        background-color: var(--137C20);

        @{b}__text {
            color: var(--137C);
        }

        @{b}__icon {
            fill: var(--137C);
        }
    }

    &_theme_green {
        background-color: var(--brandGreen20);

        @{b}__text {
            color: var(--brandGreen);
        }

        @{b}__icon {
            fill: var(--brandGreen);
        }
    }

    &_theme_red {
        background-color: var(--fury20);

        @{b}__text {
            color: var(--fury);
        }

        @{b}__icon {
            fill: var(--fury);
        }
    }

    &__text {
        .fontFamily();

        margin-left: 4px;
    }

    &__icon {
        height: 20px;
        width: 20px;
    }
}
