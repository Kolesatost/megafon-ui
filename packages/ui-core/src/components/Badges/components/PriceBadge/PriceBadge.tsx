import * as React from 'react';
import { cnCreate, filterDataAttrs } from '@megafon/ui-helpers';
import AttentionIcon from '@megafon/ui-icons/system-16-attention_16.svg';
import CheckIcon from '@megafon/ui-icons/system-16-check_invert_16.svg';
import PriceIcon from '@megafon/ui-icons/system-16-price_16.svg';
import TimerIcon from '@megafon/ui-icons/system-16-timer_16.svg';
import * as PropTypes from 'prop-types';
import './PriceBadge.less';

export const PriceBadgeTheme = {
    RED: 'red',
    GREY: 'grey',
    GREEN: 'green',
    ORANGE: 'orange',
} as const;

export const PriceBadgeIcon = {
    TIMER: 'timer',
    PRICE: 'price',
    CHECK: 'check',
    ATTENTION: 'attention',
} as const;

type PriceBadgeThemeType = typeof PriceBadgeTheme[keyof typeof PriceBadgeTheme];
type PriceBadgeIconType = typeof PriceBadgeIcon[keyof typeof PriceBadgeIcon];

const getPriceBadgeIcon = (iconType: PriceBadgeIconType) => {
    switch (iconType) {
        case PriceBadgeIcon.TIMER:
            return TimerIcon;
        case PriceBadgeIcon.PRICE:
            return PriceIcon;
        case PriceBadgeIcon.CHECK:
            return CheckIcon;
        case PriceBadgeIcon.ATTENTION:
            return AttentionIcon;
        default:
            return TimerIcon;
    }
};

export interface IPriceBadgeProps {
    /** Адаптивный режим */
    isAdaptive?: boolean;
    /** Тип иконки */
    iconType?: PriceBadgeIconType;
    /** Цветовая тема */
    theme?: PriceBadgeThemeType;
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Дополнительные data-атрибуты к внутренним элементам */
    dataAttrs?: {
        root?: Record<string, string>;
    };
    children: JSX.Element[] | Element[] | JSX.Element | Element | string;
}

const cn = cnCreate('mfui-price-badge');
const PriceBadge: React.FC<IPriceBadgeProps> = ({
    iconType = 'timer',
    isAdaptive = false,
    theme = 'grey',
    className,
    dataAttrs,
    children,
}) => {
    const Icon = getPriceBadgeIcon(iconType);

    return (
        <div {...filterDataAttrs(dataAttrs?.root)} className={cn({ theme, adaptive: isAdaptive }, className)}>
            <div className={cn('icon-container')}>
                <Icon className={cn('icon')} />
            </div>
            <span className={cn('text')}>{children}</span>
        </div>
    );
};

PriceBadge.propTypes = {
    isAdaptive: PropTypes.bool,
    iconType: PropTypes.oneOf(Object.values(PriceBadgeIcon)),
    theme: PropTypes.oneOf(Object.values(PriceBadgeTheme)),
    className: PropTypes.string,
    dataAttrs: PropTypes.shape({
        root: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
    }),
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.element.isRequired),
        PropTypes.element,
        PropTypes.string,
    ]).isRequired,
};

export default PriceBadge;
