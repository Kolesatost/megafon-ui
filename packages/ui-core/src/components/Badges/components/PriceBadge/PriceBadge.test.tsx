import * as React from 'react';
import { shallow } from 'enzyme';
import PriceBadge from './PriceBadge';

const defaultChildren = 'badge text';

describe('<PriceBadge />', () => {
    describe('snapshots', () => {
        it('renders with default props', () => {
            const wrapper = shallow(<PriceBadge>{defaultChildren}</PriceBadge>);

            expect(wrapper).toMatchSnapshot();
        });

        it('renders with all props', () => {
            const wrapper = shallow(
                <PriceBadge iconType="price" theme="green" isAdaptive>
                    {defaultChildren}
                </PriceBadge>,
            );

            expect(wrapper).toMatchSnapshot();
        });

        it('renders with check icon', () => {
            const wrapper = shallow(<PriceBadge iconType="check">{defaultChildren}</PriceBadge>);

            expect(wrapper).toMatchSnapshot();
        });

        it('renders with attention icon', () => {
            const wrapper = shallow(<PriceBadge iconType="attention">{defaultChildren}</PriceBadge>);

            expect(wrapper).toMatchSnapshot();
        });
    });
});
