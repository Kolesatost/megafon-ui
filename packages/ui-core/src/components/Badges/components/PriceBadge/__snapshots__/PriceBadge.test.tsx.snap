// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PriceBadge /> snapshots renders with all props 1`] = `
<div
  className="mfui-price-badge mfui-price-badge_theme_green mfui-price-badge_adaptive"
>
  <test-file-stub
    className="mfui-price-badge__icon"
  />
  <span
    className="mfui-price-badge__text"
  >
    badge text
  </span>
</div>
`;

exports[`<PriceBadge /> snapshots renders with attention icon 1`] = `
<div
  className="mfui-price-badge mfui-price-badge_theme_grey"
>
  <test-file-stub
    className="mfui-price-badge__icon"
  />
  <span
    className="mfui-price-badge__text"
  >
    badge text
  </span>
</div>
`;

exports[`<PriceBadge /> snapshots renders with check icon 1`] = `
<div
  className="mfui-price-badge mfui-price-badge_theme_grey"
>
  <test-file-stub
    className="mfui-price-badge__icon"
  />
  <span
    className="mfui-price-badge__text"
  >
    badge text
  </span>
</div>
`;

exports[`<PriceBadge /> snapshots renders with default props 1`] = `
<div
  className="mfui-price-badge mfui-price-badge_theme_grey"
>
  <test-file-stub
    className="mfui-price-badge__icon"
  />
  <span
    className="mfui-price-badge__text"
  >
    badge text
  </span>
</div>
`;
